pkg/translator/zipkin/zipkinv1/json.go

```diff
@@ -106,215 +97,151 @@ type binaryAnnotation struct {
-		// Span kind has no equivalent in OC, so we cannot represent it in the Kind field.
-		// We will set a TagSpanKind attribute in the span. This will successfully transfer
-		// in the pipeline until it reaches the exporter which is responsible for
-		// reverse translation.
-		if ocSpan.Attributes == nil {
-			ocSpan.Attributes = &tracepb.Span_Attributes{}
-		}
-		if ocSpan.Attributes.AttributeMap == nil {
-			ocSpan.Attributes.AttributeMap = make(map[string]*tracepb.AttributeValue, 1)
-		}
-		ocSpan.Attributes.AttributeMap[tracetranslator.TagSpanKind] =
-			&tracepb.AttributeValue{Value: &tracepb.AttributeValue_StringValue{
-				StringValue: &tracepb.TruncatableString{Value: string(extendedKind)},
-			}}
-	}
-}
-
-func zipkinV1BinAnnotationsToOCAttributes(binAnnotations []*binaryAnnotation, parseStringTags bool) (attributes *tracepb.Span_Attributes, status *tracepb.Status, fallbackServiceName string) {
+func zipkinV1BinAnnotationsToOCAttributes(span ptrace.Span, binAnnotations []*binaryAnnotation, parseStringTags bool) string {
+	var fallbackServiceName string
 	if len(binAnnotations) == 0 {
-		return nil, nil, ""
+		return fallbackServiceName
 	}
 
 	sMapper := &statusMapper{}
 	var localComponent string
-	attributeMap := make(map[string]*tracepb.AttributeValue)
 	for _, binAnnotation := range binAnnotations {
 
 		if binAnnotation.Endpoint != nil && binAnnotation.Endpoint.ServiceName != "" {
 			fallbackServiceName = binAnnotation.Endpoint.ServiceName
 		}
 
-		pbAttrib := parseAnnotationValue(binAnnotation.Value, parseStringTags)
-
 		key := binAnnotation.Key
-
 		if key == zipkincore.LOCAL_COMPONENT {
 			// TODO: (@pjanotti) add reference to OpenTracing and change related tags to use them
 			key = "component"
 			localComponent = binAnnotation.Value
 		}
 
-		if drop := sMapper.fromAttribute(key, pbAttrib); drop {
+		val := parseAnnotationValue(binAnnotation.Value, parseStringTags)
+		if drop := sMapper.fromAttribute(key, val); drop {
 			continue
 		}
 
-		attributeMap[key] = pbAttrib
-	}
-
-	status = sMapper.ocStatus()
-
-	if len(attributeMap) == 0 {
-		return nil, status, ""
+		val.CopyTo(span.Attributes().PutEmpty(key))
 	}
 
 	if fallbackServiceName == "" {
 		fallbackServiceName = localComponent
 	}
 
-	attributes = &tracepb.Span_Attributes{
-		AttributeMap: attributeMap,
-	}
-
-	return attributes, status, fallbackServiceName
+	sMapper.ocStatus(span.Status())
+	return fallbackServiceName
 }
 
-func parseAnnotationValue(value string, parseStringTags bool) *tracepb.AttributeValue {
-	pbAttrib := &tracepb.AttributeValue{}
+func parseAnnotationValue(value string, parseStringTags bool) pcommon.Value {
+	val := pcommon.NewValueEmpty()
 
 	if parseStringTags {
 		switch zipkin.DetermineValueType(value) {
 		case pcommon.ValueTypeInt:
 			iValue, _ := strconv.ParseInt(value, 10, 64)
-			pbAttrib.Value = &tracepb.AttributeValue_IntValue{IntValue: iValue}
+			val.SetInt(iValue)
```

Comment:
Not a big deal but curious why not `return pcommon.NewValueInt(iValue)`?

Reply:
Done.